xds/src/main/java/io/grpc/xds/Bootstrapper.java

```diff
@@ -125,9 +120,163 @@ private static Bootstrap readConfig() throws IOException {
   }
 
   @VisibleForTesting
-  static Bootstrap parseConfig(String rawData) throws InvalidProtocolBufferException {
-    Bootstrap.Builder bootstrapBuilder = Bootstrap.newBuilder();
-    JsonFormat.parser().merge(rawData, bootstrapBuilder);
-    return bootstrapBuilder.build();
+  @SuppressWarnings("unchecked")
+  static BootstrapInfo parseConfig(String rawData) throws IOException {
+    Map<String, ?> rawBootstrap = (Map<String, ?>) JsonParser.parse(rawData);
+
+    Map<String, ?> rawServerConfig = (Map<String, ?>) rawBootstrap.get("xds_server");
+    if (rawServerConfig == null) {
+      throw new IOException("Invalid bootstrap: 'xds_server' does not exist.");
+    }
+    // Field "server_uri" is required.
+    String serverUri = (String) rawServerConfig.get("server_uri");
+    if (serverUri == null) {
+      throw new IOException("Invalid bootstrap: 'xds_server : server_uri' does not exist.");
+    }
+    List<ChannelCreds> channelCredsOptions = new ArrayList<>();
+    List<?> rawChannelCredsList = (List<?>) rawServerConfig.get("channel_creds");
+    // List of channel creds is optional.
+    if (rawChannelCredsList != null) {
+      for (Object channelCreds : rawChannelCredsList) {
+        Map<String, ?> channelCredsItem = (Map<String, ?>) channelCreds;
+        String type = (String) channelCredsItem.get("type");
+        if (type == null) {
+          throw new IOException("Invalid bootstrap: 'channel_creds' contains unknown type.");
+        }
+        ChannelCreds creds = new ChannelCreds(type);
+        if (channelCredsItem.containsKey("config")) {
+          creds.config = (Map<String, ?>) channelCredsItem.get("config");
+        }
+        channelCredsOptions.add(creds);
+      }
+    }
+    ServerConfig serverConfig = new ServerConfig(serverUri, channelCredsOptions);
+
+    Map<String, ?> rawNode = (Map<String, ?>) rawBootstrap.get("node");
+    if (rawNode == null) {
+      throw new IOException("Invalid bootstrap: 'node' does not exist.");
+    }
+    // Fields in "node" are not checked.
+    Node.Builder nodeBuilder = Node.newBuilder();
+    String id = (String) rawNode.get("id");
+    if (id != null) {
+      nodeBuilder.setId(id);
+    }
+    String cluster = (String) rawNode.get("cluster");
+    if (cluster != null) {
+      nodeBuilder.setCluster(cluster);
+    }
+    Map<String, ?> metadata = (Map<String, ?>) rawNode.get("metadata");
+    if (metadata != null) {
+      Struct.Builder structBuilder = Struct.newBuilder();
+      for (Map.Entry<String, ?> entry : metadata.entrySet()) {
+        structBuilder.putFields(entry.getKey(), convertToValue(entry.getValue()));
+      }
+      nodeBuilder.setMetadata(structBuilder);
```

Comment:
`nodeBuilder.setMetadata(convertToValue(metadata))`

Reply:
This is subtle to explain...

A `Struct` is a JSON __object__, aka, key-value pairs. Each value is a `Value`.
A `Value` is anything that is valid for JSON field value, which can be __string__, __object__, __boolean__, etc. 

So a `Struct` contains `Value` and a `Value` may be (actually "contains" in terms of protobuf definition) a `Struct`.

Here `setMetadata()` requires `Struct`.